### DISCUSSION TITLE
page form tree

### DISCUSSION THREAD

**Discussion by cole-dda** [pymupdf/PyMuPDF#2390] [Q&A]
**Is your feature request related to a problem? Please describe.**
page can include form，get from tree

**Describe the solution you'd like**
forms = page.get_forms()
forms=[{
 name:'form1',
children:[{name:'form2'},{name:'form3'}]
}]

objs = page.get_texttrace()
any object include "form" key

**Comment by JorjMcKie** [comment]
Your description is very vague unfortunately, and also seems to be inconsistent.
1. You already can access a page's widgets e.g., via the iterator `page.widgets()`. This yields the fields one by one.
2. Mentioning method `page.get_texttrace()` in context of form fields is confusing, because it delivers low-level text information only, no widgets. Please see the documentation.

Let us transfer this item to the "Discussions" tab for further clarification.

**Comment by JorjMcKie** [comment]
Assuming that you are referring widget hierarchies within a page:

You do have access to the xref of widgets in the page, `field.xref`. Using document method `xref_get_key()`, you can access PDF dictionary items of the field's object definition, for example `kids = doc.xref_get_key(field.xref, "Kids")`, which returns `("array", "[n 0 R m 0 R ...]")`, an array of xref numbers of the children of that field.
Each child can then be loaded via `kid_field = page.load_widget(n)`.

Similarly for the parent of a field `p_tuple = doc.xref_get_key(field.xref, "Parent")`. The result in this case should be `("xref", "n 0 R")` if there in fact is a parent.
